### ISSUE TITLE
Multi-Site Support for SLO

### ISSUE THREAD

**Issue by dsmrt** [flipboxfactory/saml-sp#72]
With Auth0 for example, you can manage multiple SPs under one configuration. Login seems to work fine due to us setting the Assertion Consumer Service URL in the AuthnRequest payload but there isn't an equivalent directive in the logout request, so we may need to get creative with RelayState for SLO. Evaluate error thrown when there's not session as well.

Example of the issue:

Craft has 2 sites: site-a.com and site-b.com . The SAML configuration is setup for site-a.com. A user logs-in to site-b.com. Then when logging-out, the request comes from site-b.com the request is sent to Auth0 with the correct info, then Auth0 sends the response back to site-a.com (instead of site-b.com), where the user has no session.

**Comment by dsmrt** [comment]
Might look at overloading `RelayState` for this. Possibly a base64'd json object that contains SLO host origin and url redirect.

**Comment by smaida** [comment]
@dsmrt Curious if you have a sense of timing on this one?

**Comment by dsmrt** [comment]
Hey @smaida  ... Sorry to leave you hanging here. I was hoping to get to this quicker but a couple other issues came up and I'm not crazy about the solution for this. 

I'll re-evaluate this and see what I can come up with. I may add this with a feature flag (in settings), so it'll be something you have to turn on to enable the feature. 

With all of that said, I'll try and update this ticket by Tuesday (9/8/2020) with a patch or, at the least, a plan.

**Comment by dsmrt** [comment]
Remember, if you do need to get this to work sooner, you can setup one SP/Provider (on craft) per craft site and configure that appropriately with the IdP.

**Comment by dsmrt** [comment]
Hey @smaida, 

Here's where I'm at. 

I started down the road of having events and relay state context be available but I would still require implementers (you or anyone else) to add custom code to handle things as needed based on your scenario. So I back-peddled. 

Out of the box, there's no easy way to do this without things getting a little messy. Each IdP can have a slightly different way of handling Single Log Out (SLO) and I'm worried that complicating this specifically for your configuration or Auth0's will cause issues for others. 

Basically, the best and cleanest way of handling multi-site configurations, is to configure them separately, as separate sites/providers.

**Comment by adamtsai1** [comment]
@dsmrt Hi Damien, to follow up on your suggestion of adding multiple SP's, I am having trouble getting that to work. When I try to add a record, the Entity ID reverts back to the Default Entity ID (set in settings) no matter what I enter. Assuming each SP record is supposed to correspond to a site in my multi-site config (e.g. "mywebsite.test" vs "mywebsite.ca"), how am I supposed to set that up?

**Comment by adamtsai1** [comment]
Just for context, I am using saml-sp version 2.5.1.

**Comment by dsmrt** [comment]
Are you modifying the settings here? https://saml-sp.flipboxfactory.com/configure/settings.html#entityid

You can put logic in that setting to match based on site URL, environmental variable, or other environment configuration. If you can set environmental variables based on site, you can use the backend setting (via the Plugin control  panel settings) and set the environmental variable there. 

You might be able to do something as simple as this: 
Within the `index.php` (web entrypoint) or within the `config/saml-sp.php` ....
```php
putenv("ENTITY_ID=$siteHost");
```

Then you can set that in the backend plugin cp setting or within the settings config (`config/saml-sp.php`) file like so:

```php
return [
   'entityId' => '$ENTITY_ID',
]; 
```

**Note:**
I just updated saml-core to 3.1.1 which fixes https://github.com/flipboxfactory/saml-sp/issues/85 which is a bug in the control panel settings. Run a `composer update flipboxfactory/saml-core` to patch that.

**Comment by adamtsai1** [comment]
@dsmrt Thanks, that seems to get me a little closer; however, while I am able to create a SP for each site with a corresponding entity ID, it appears that the auto-generated ACS and SLO bindings are all pointing at the base site. I assume that is generating off of a configuration somewhere, but is there any way to have it be dynamic by host domain (or entity) as well?

**Comment by dsmrt** [comment]
Just to understand better, is the control panel url the same as the site url/base url you are attempting to create the SP for?

For example, you have 2 sites and the cp and site base url look like so 
1. cpUrl: site1.com/admin,  baseUrl: site1.com
2. cpUrl: site2.com/admin,  baseUrl: site2.com

There might be some configurations I have to test here.

**Comment by adamtsai1** [comment]
Yes, I just tried it again to double-check. I logged in to site 2's CP and created a service provider, which allowed the Entity ID to match the site 2 domain. But the ACS and SLO values are using site 1's URL.

**Comment by dsmrt** [comment]
Cool. Thanks for clarifying. Seems like there's some work to do with selecting the site correctly. If you need to get this working quickly, you can update the metadata in the db for now while I work on a proper fix, or possibly, update/add the correct configuration (entityId, ACS and SLO endpoints, etc) to the IdP. The auth flow may work without needing to update the metadata if the IdP has the correct info.

Seems like this issue is expanding into more multi-site issues in general. It would be nice to make multi-site management easier in general. I think I have some ideas, let me get back to you. Thanks for your patience!

**Comment by dsmrt** [comment]
@adamtsai1 FYI, I added this issue https://github.com/flipboxfactory/saml-sp/issues/89. 

There's a lot of moving parts here so I don't timeline, although, I'd like to get this out in the next month or so. Hope this helps!

**Comment by dsmrt** [comment]
Sorry to take so long with this but I think I have something working well with the following release: https://github.com/flipboxfactory/saml-sp/releases/tag/2.6.0

I did change the way endpoints work, and they are based on Craft Sites, so to take advantage of these features:
1. update of course
2. Re-save "My provider" (to recreate the metadata)
3. Sync it with your metadata with your IdP

Finally, Please test this with a dev environment. Let me know if you need help with this!
